src/anyp/Uri.cc

```diff
@@ -514,27 +515,37 @@ AnyP::Uri::absolute() const
  *        After copying it on in the first place! Would be less code to merge the two with a flag parameter.
  *        and never copy the query-string part in the first place
  */
-char *
+SBuf
 urlCanonicalCleanWithoutRequest(const SBuf &url, const HttpRequestMethod &method, const AnyP::UriScheme &scheme)
 {
-    LOCAL_ARRAY(char, buf, MAX_URL);
-
-    snprintf(buf, sizeof(buf), SQUIDSBUFPH, SQUIDSBUFPRINT(url));
-    buf[sizeof(buf)-1] = '\0';
-
     // URN, CONNECT method, and non-stripped URIs can go straight out
+    SBuf out = url;
     if (Config.onoff.strip_query_terms && !(method == Http::METHOD_CONNECT || scheme == AnyP::PROTO_URN)) {
-        // strip anything AFTER a question-mark
-        // leaving the '?' in place
-        if (auto t = strchr(buf, '?')) {
-            *(++t) = '\0';
-        }
+        // strip anything after a path segment
+        // leaving the delimiter in place as a visual cue of truncation (if any)
+        static const CharacterSet pathDelims("path-delim", "?#");
+        const auto pathEnd = url.findFirstOf(pathDelims);
+        if (pathEnd != SBuf::npos)
+            out = url.substr(0, pathEnd+1);
     }
 
-    if (stringHasCntl(buf))
-        xstrncpy(buf, rfc1738_escape_unescaped(buf), MAX_URL);
+    // check for unescaped control characters
+    // XXX: emulates old stringHasCtl() which matched true
+    //      only for the UTF-8 characters 0x80-09f.
+    // TODO: RFC 1738, 5234 and 7230 require the full UTF-8
+    //      range (0x80-0xFF) be encoded along with
+    //      the ASCII control characters 0x01-0x1F, 0x7f
+    //      and NUL (0x00)
```

Comment:
Most of the range(s) required are already there. As you saw the handling of NUL is (still) buggy. The TODO and XXX were about adding the missing ones and possibly fixing that NUL handling. Since that is a visible behaviour change it is left out of scope for this PR.

I have re-written the comments. I hope the new text is clear enough.